### DISCUSSION TITLE
Question re: drawing addon - correct way to clear view/scene/backend?

### DISCUSSION THREAD

**Discussion by bandtank** [mozman/ezdxf#947] [Q&A] [answered]
Using the `drawing` addon, is there a built-in way to clear the scene/reset the view/unload the DXF file? When the example first loads without a CAD file, the scene is white and uninitialized. How would I return it to that state?

So far, I am using `self._cad._backend.clear()`, which seems to mostly work. However, the scene appears to still be loaded because the background is black instead of white, so I'm not sure if that is sufficient to fully reset the frontend and backend. In other words, I'm trying to figure out how to close a DXF file and return the program to the initial state.

**Comment by mozman** [comment, marked as answer]
You can follow the [qtviewer.py](https://github.com/mozman/ezdxf/blob/master/src/ezdxf/addons/drawing/qtviewer.py) example. Start at the `CADWidget.set_document()` method.

https://github.com/mozman/ezdxf/blob/860dd34f7cc7747f088dc47b32a5dc1ff1c9c3bd/src/ezdxf/addons/drawing/qtviewer.py#L259

The viewer creates a new backend for each new loaded DXF document. This seems to be the safest way to do it. 

The  `CADWidget.draw_layout()` method draws another layout of the same DXF document without creating a new backend, which also seems to work. 

https://github.com/mozman/ezdxf/blob/860dd34f7cc7747f088dc47b32a5dc1ff1c9c3bd/src/ezdxf/addons/drawing/qtviewer.py#L287

**Reply by bandtank** [reply]
I added this to the CAD Widget:
```
  def reset(self):
    self._view.clear()
    self._bbox_cache = ezdxf.bbox.Cache()
    self._doc = None
    self._render_context = None
    self._visible_layers = set()
    self._current_layout = None
    self._reset_backend()
```
The goal was to set as many attributes in the widget as possible to the initial states. It is called from the CAD viewer like this:
```
  def unloadDXFFile(self):
    self._cad.reset()
    self._doc = None
```
However, the screen was still black instead of white. I then added the following code to the CAD Graphics View:
```
  def clear(self):
    self.setScene(qw.QGraphicsScene())
```
Now the screen is white. It seems like the replacement of the scene fixes the graphical aspect while everything else resets the internal state.

Thanks for the help.
